test/test_forbidden.py

```diff
@@ -225,3 +225,36 @@ def test_and_conjunction(self):
             self.assertEqual(results[i], is_forbidden)
 
             self.assertFalse(total_and.is_forbidden({}, strict=False))
+
+    def test_relation(self):
+        hp1 = CategoricalHyperparameter("cat_int", [0, 1])
+        hp2 = UniformIntegerHyperparameter("int", 0, 2)
+        hp3 = UniformIntegerHyperparameter("int2", 0, 2)
+        hp4 = UniformFloatHyperparameter("float", 0, 2)
+        hp5 = CategoricalHyperparameter("str", ['foo', 'bar'])
```

Comment:
Seeing this, I'm wondering whether we should forbid categorical and ordinal hyperparameters for now or add documentation that the forbiddens can result in weird behavior here? Or we'd need to think how to implement less than and greater than correctly for ordinals and categorical (as they can behave differently when they are compared with their transformed version (_is_forbidden_vector) than their untransformed version (_is_forbidden)).

Reply:
I specifically require the option to compare a list of integers with an int hyperparameter. Therefore, my idea was to check the compatibility when evaluating the actual configuration. In case two hyperparameters are not comparable python raises a TypeError. That's the reason why I check for a TypeError in the test case below when comparing `1 == 'foo'`.

Regarding the problem with `_is_forbidden_vector`: I explicitly convert from the vector representation to the represented value in line #580 to circumvent this problem. So basically `_is_forbidden_vector` behaves like `_is_forbidden`.

If you require to validate the configuration space during construction, I can adapt the code to only accept NumericalHyperparamter.

Reply:
Okay, got it. So the behavior basically is:
* int and float: all fine
* categorical or ordinal that contains int or float: all fine
* categorical or ordinal that contain strings: ordering is determined based on a string comparison, not on how the hyperparameter type should treat it (categorical should always return false for larger or smaller, ordinal should consider the ordinal relation).
If that's the case it would be great if you could add this to the docstring so that users aren't surprised.

Reply:
Thanks a lot for the explanation and the updated tests, that makes sense. Would you mind adding a case for Ordinal that covers a hyperparameter like `OrdinalHyperparameter("water_temperature", ["cold", "luke-warm", "hot", "boiling"])` to test that we're failing or comparing according to string comparisons, not the actual comparisons?